src/patterns.jl

```diff
@@ -118,29 +124,45 @@ For use with [`HessianTracer`](@ref).
 
 ## Expected interface
 
-* `myempty(::Type{MyPattern})`: return a pattern representing a new number (usually an empty pattern)
-* `seed(::Type{MyPattern}, i::Integer)`: return an pattern that only contains the given index `i` in the first-order representation
+* [`myempty`](@ref)
+* [`create_patterns`](@ref)
 * `gradient(p::MyPattern)`: return non-zero indices `i` in the first-order representation
 * `hessian(p::MyPattern)`: return non-zero indices `(i, j)` in the second-order representation
+* [`isshared`](@ref) in case the pattern is shared (mutates). Defaults to false.
 """
 abstract type AbstractHessianPattern <: AbstractPattern end
 
 """
-    IndexSetHessianPattern(vector::AbstractGradientPattern, mat::AbstractMatrixPattern)
+$(TYPEDEF)
+
+Hessian sparsity pattern represented by:
+* an `AbstractSet` of indices ``i`` of non-zero values representing first-order sparsity
+* an `AbstractSet` of index tuples ``(i,j)`` of non-zero values representing second-order sparsity
+
+## Fields
+$(TYPEDFIELDS)
+
+## Internals
 
-Gradient and Hessian sparsity patterns constructed by combining two AbstractSets.
+The last type parameter `shared` is a `Bool` indicating whether the `hessian` field of this object should be shared among all intermediate scalar quantities involved in a function.
 """
-struct IndexSetHessianPattern{I<:Integer,SG<:AbstractSet{I},SH<:AbstractSet{Tuple{I,I}}} <:
-       AbstractHessianPattern
+struct IndexSetHessianPattern{
+    I<:Integer,SG<:AbstractSet{I},SH<:AbstractSet{Tuple{I,I}},mutating
```

Comment:
I prefer `shared` to `mutating`, see above

Reply:
Its defining property is that it is shared between indices

Reply:
Theoretically there could be a non mutating version which just copies it, if copies were free

Reply:
I chose this because I feel like mutation is a much more common concept for Julia devs.
Walther's approach mutates values stored at a shared pointer.

And technically speaking, in this PR, pointers aren't *always* shared, see the `myempty` discussion above and https://github.com/adrhill/SparseConnectivityTracer.jl/pull/135#discussion_r1650967036. What this PR does is **allow** operators to mutate their tracer arguments.

Reply:
I renamed it, but I hope you see that mutating operators are the defining concept of this PR.